### PR TITLE
Migration of input collection parameters to untracked in DQM/BeamMonitor package

### PR DESCRIPTION
**Pull request by NiharSaha**
#### PR description:
This PR has switched all the tracked parameters of DQM/BeamMonitor to untracked parameters to read input collections which is followed by the discussion: https://github.com/cms-sw/cmssw/pull/36998#discussion_r815286443

#### PR validation:
Tested in CMSSW_12_4_X via runTheMatrix.py -l limited -i all --ibeos

#### if this PR is a backport please specify the original PR and why you need to backport that PR:
NA

### CODE REVIEW

**Comment by tvami** [comment]
@cmsbuild , please test

**Comment by tvami** [comment]
My understanding is that if the module has a `fillDescription` than the python configs dont need explicit types, and actually it's advised not to have them at all.  

i.e. 
```
dxBin = cms.untracked.int32(200),
```
could become
```
dxBin = 200,
```

Although, I have to say that I didnt check if that given module has a `fillDescription` or not.

**Comment by tvami** [comment]
I see that `PixelVTXMonitor` is still a legacy module, that should be migrated, I guess for a follow-up PR:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-b75ee9/24507/deprecated-warnings.log

**Comment by tvami** [comment]
@cmsbuild , please test

**Comment by mmusich** [comment]
> This PR has switched all the tracked parameters of DQM/BeamMonitor to untracked parameters to read input collections 

1) here *all* parameters have been made untracked, not only the ones related to read input collections.

2) what's the advantage of doing so?

**Comment by emanueleusai** [comment]
@mmusich Nihar is following the decision/prescription we took here: https://github.com/cms-sw/cmssw/pull/36998#discussion_r815286443

yes perhaps we should adjust the PR so that only the input collections are changed to untracked.

**Comment by NiharSaha** [comment]
> My understanding is that if the module has a `fillDescription` than the python configs dont need explicit types, and actually it's advised not to have them at all.
> 
> i.e.
> 
> ```
> dxBin = cms.untracked.int32(200),
> ```
> 
> could become
> 
> ```
> dxBin = 200,
> ```
> 
> Although, I have to say that I didnt check if that given module has a `fillDescription` or not.

Okay, let me try this out. Thank you for pointing this out.

**Comment by mmusich** [comment]
> following the decision/prescription we took here: https://github.com/cms-sw/cmssw/pull/36998#discussion_r815286443

still not sure what's the benefit (the cost is a lot of code churn)

> yes perhaps we should adjust the PR so that only the input collections are changed to untracked.

By all means, at least do that.

**Comment by perrotta** [comment]
> My understanding is that if the module has a `fillDescription` than the python configs dont need explicit types, and actually it's advised not to have them at all.
> 
> i.e.
> 
> ```
> dxBin = cms.untracked.int32(200),
> ```
> 
> could become
> 
> ```
> dxBin = 200,
> ```
> 
> Although, I have to say that I didnt check if that given module has a `fillDescription` or not.

Well. it is recommended not to write the explicit types for parameters modified when you `clone` an existing module, because it is a safer way to redefine them (see for example item 4 in https://twiki.cern.ch/twiki/bin/view/CMS/OpenRecoTasks#Review_and_optimization_of_the_p). This is independent from the fact that the code used to define that module has a fillDescription method or not.

You cannot avoid writing the explicit type if you define a module anew. Also this is independent from the existence of a fillDescription method in the C++ code that defines that module.

Having a fillDescription method is needed to validate the parameters that you insert in the configuration, and fill with a default the ones that are not explicitely defined in the configuration.

Of course, the fillDescription method also automatically provides you with a default configuration file from which you can always clone your module, thus avoiding to write the explicit types for the modified parameters in the cloned configuration.

**Comment by tvami** [comment]
Thanks Andrea for the clarification

> Of course, the fillDescription method also automatically provides you with a default configuration file from which you can always clone your module, thus avoiding to write the explicit types for the modified parameters in the cloned configuration.

And this is the centrally recommended way to go, right?

**Comment by emanueleusai** [comment]
> still not sure what's the benefit (the cost is a lot of code churn)

@mmusich the original concern about the input collection parameters was from @makortel and @fwyzard so I'll let them comment on the benefits of standardizing the parameter type. 

btw the term"code churn" is not used correctly here. It's just a large number of small (relatively simple) changes.

**Comment by mmusich** [comment]
> btw the term"code churn" is not used correctly here. It's just a large number of small (relatively simple) changes.

mmmh, if you plan to embark in changing the "trackedness" of all the parameters for DQM modules in `cmssw`, will you also take care of updating the standalone configuration files which might be using such modules but are either not used in central sequences and / or unit tests?

**Comment by emanueleusai** [comment]
I stand by the recommendation from @fwyzard and @makortel from the original thread. If that's what needed to make you happy we will track the cfgs and modify those as well.

If you need a more appropriate space to voice and address your concerns, I am more than happy to schedule another "round table" discussion at the next DQM General meeting. Discussing relevant or controversial PRs is well within the scope of the meeting. We can invite Matti, Andrea, and @jfernan2 (as former DQM convener who handed over this parameter conversion task to me earlier this year) to make sure everybody is happy with these modifications. Let me know or feel free to reach out to me privately.

**Comment by mmusich** [comment]
> If you need a more appropriate space to voice and address your concerns, I am more than happy to schedule another "round table" discussion at the next DQM General meeting. 

I am not *a priori* against this change, provided is done carefully (e.g. applying only to the input collections as the original suggestion and not indiscriminately to all parameters plus tracking down all dependent cfgs), but I would like just to understand better the purpose. After all we've been running like this for 10+ years at this point, so if it's just for aesthetics reasons one might wonder if it's worth the effort (and the possible related breakage of untested code).

**Comment by tvami** [comment]
Hi All, what's the status of this PR? @emanueleusai was there a discussion about this in the DQM meeting?

@NiharSaha independently of possible out changes as the outcome of the DQM meeting, you for sure need to rebase to the latest IB, the PR currently "has conflicts that must be resolved"

**Comment by emanueleusai** [comment]
We will continue with the PR, I think Nihar is busy with other matters atm.
We will rebase.

**Comment by tvami** [comment]
What's the plan with this PR? There hasnt been an update for a long period of time, should it be just closed?

**Comment by emanueleusai** [comment]
@NiharSaha do you have any update?

**Comment by NiharSaha** [comment]
> What's the plan with this PR? There hasnt been an update for a long period of time, should it be just closed?

Hi @tvami, extremely sorry for not updating for long. I am figuring out the parameters to change and will rebase the same soon.

**Comment by NiharSaha** [comment]
> @NiharSaha do you have any update?

Hi @emanueleusai , I am changing the parameters as discussed in this thread and will rebase the PR soon.

**Comment by NiharSaha** [comment]
Hi, I need to update something here about the CMSSW release. As the parent CMSSW version where the commit was made is no longer available, hence I am planning to make a new PR with the latest CMSSW version, closing the current PR. I hope that is okay? Also, while keeping in mind that I need to change only input collection parameters, I am considering the variables/parameters read with <InputTag> as input collection parameters for now. Apart from that, are there any other parameters that are considered input collection parameters? Please let me know so that I can follow the same rule for all the DQM*/packages. Thank you!

**Comment by tvami** [comment]
Well you could rebase and force push to the same branch, you dont really need a new PR

**Comment by NiharSaha** [comment]
> Well you could rebase and force push to the same branch, you dont really need a new PR

Hi @tvami, when I am changing in the same CMSSW locally, no commands are working as the CMSSW release is no longer available. Do you mean that after changing everything in the latest CMSSW, I can push in this same branch (older CMSSW version)?

**Comment by tvami** [comment]
@NiharSaha here is what you need to do
```
cmsrel CMSSW_12_5_X_2022-06-16-2300
cd CMSSW_12_5_X_2022-06-16-2300/src
cmsenv
git cms-merge-topic 36273 
scram b -j
````
do your changes
```
git push git@github.com:NiharSaha/cmssw.git +HEAD:CMSSW_12_4_X_DQM_BeamMonitor
```

Please note the IBs are alive for about one week, so you'll need to do your changes in that timescale...

**Comment by NiharSaha** [comment]
> @NiharSaha here is what you need to do
> 
> ```
> cmsrel CMSSW_12_5_X_2022-06-16-2300
> cd CMSSW_12_5_X_2022-06-16-2300/src
> cmsenv
> git cms-merge-topic 36273 
> scram b -j
> ```
> 
> do your changes
> 
> ```
> git push git@github.com:NiharSaha/cmssw.git +HEAD:CMSSW_12_4_X_DQM_BeamMonitor
> ```
> 
> Please note the IBs are alive for about one week, so you'll need to do your changes in that timescale...

Thank you. It will definitely help!

**Comment by tvami** [comment]
@NiharSaha 
you need to do code-checks, and actually, after doing that please squash the 3 commits into one

**Comment by NiharSaha** [comment]
@tvami  Yes. I am not done with the required changes. After all the commits I will squash them into a single one. I just tried to resolve the conflicts with the other PRs, and got this error. I just kept as same as the master branch.

**Comment by mmusich** [comment]
test parameters:
   * addpkg = DQM/Integration

**Comment by tvami** [comment]
@cmsbuild , please test

**Comment by NiharSaha** [comment]
Hi @mmusich, I can see that I have changed to untracked in the python configuration as well for `hltResults`. Could you please help me to understand the error?

**Comment by mmusich** [comment]
> Could you please help me to understand the error?

you changed the c++ source for `BeamMonitor`, but apparently not all configuration files using it (e.g. the one of the beamspot DQM online client)

**Comment by NiharSaha** [comment]
Oh, I see! I missed that. I will modify the PR descriptions to add the dependent configurations. But, is there an effective way to know all the dependent configurations which I may not know. It will help to follow all the other packages.

**Comment by perrotta** [comment]
> Oh, I see! I missed that. I will modify the PR descriptions to add the dependent configurations. But, is there an effective way to know all the dependent configurations which I may not know? It will help to follow all the other packages.

There are five python configs with instantiations of the "BeamMonitor" DQMEDAnalyzer, see e.g.:
https://cmssdt.cern.ch/lxr/search?%21v=CMSSW_12_5_X_2022-07-03-2300&_filestring=python&_string=%22BeamMonitor%22

**Comment by NiharSaha** [comment]
Ahh..thanks a lot for this tool. Let me check and do the needful. I hope everything gets clear in this PR before moving to other packages.

**Comment by tvami** [comment]
@cmsbuild , please test

**Comment by tvami** [comment]
+db
 * tests pass
 * the usage of untested code is tracked down with lxr, and the modification are done there too

**Comment by mmusich** [comment]
@NiharSaha did you check for other configuration files containing instances of the other analyzer codes you have changed?

**Comment by NiharSaha** [comment]
Hi @mmusich, I have checked for all the instances in all the configuration files for this package using LXR. There are four configuration files newly added to the new commit where the instances are present. There are two parameters viz `hltResults` and `tcdsRecord` which are present in two and four configuration files respectively. All the other parameters appear single times in configuration files. 

In summary, there are 6 input collection parameters which are appeared in 7 configuration files for BeamMonitor.

**Comment by mmusich** [comment]
@NiharSaha 

> In summary, there are 6 input collection parameters which are appeared in 7 configuration files for BeamMonitor.

what about `BeamSpotDipServer` and `PixelVTXMonitor` ?

**Comment by NiharSaha** [comment]
Yeah, I have checked that too. There is only one parameter in `BeamSpotDipServer` that appeared in the `BeamSpotDipServer_cff.py `only and there are 3 parameters in `PixelVTXMonitor`, all appeared in `PixelVTXMonitor_cfi.py` using LXR.

**Comment by emanueleusai** [comment]
+1

**Comment by perrotta** [comment]
+1
- The "+1"'s make me understand that the migration was finally approved
- It also seems to work in the tested workflows
